### PR TITLE
Separate profiling tests from p2p tests

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56412 Separate profiling tests from p2p tests**

We are investigating some flaky profiiling tests such as https://github.com/pytorch/pytorch/issues/56146. One issue is that the profiling tests are tightly coupled to these send/recv tests, hence if this test is disabled, we lose signal round send/recv collectives tests.

To mitigate this, separate the tests into ones that only test send/recv, and ones that test it with profiling. This way flakiness should not result in the send/recv only tests being disabled.

Differential Revision: [D27864845](https://our.internmc.facebook.com/intern/diff/D27864845/)

### CODE REVIEW

File: torch/testing/_internal/distributed/distributed_test.py [hunk 1]
```diff
@@ -942,12 +942,16 @@ def test_send_recv_nccl(self):
...
 
         # SEND RECV
-        @unittest.skipIf(BACKEND == "nccl", "Nccl does not support send/recv")
-        def test_send_recv(self):
+        def _test_send_recv(self, enable_profiling):
             rank = dist.get_rank()
             send_size = rank + 1
             tensor = _build_tensor(send_size)
-            with torch.autograd.profiler.profile(record_shapes=True) as prof:
+            profile_ctx = (
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/distributed_test.py:949]: this occurs multiple times, shall we dedup this?

File: torch/testing/_internal/distributed/distributed_test.py [hunk 2]
```diff
@@ -1021,39 +1036,56 @@ def test_send_recv_any_source(self):
...
+                global_recv_ranks_list = []
+                for tensor in global_recv_ranks:
+                    global_recv_ranks_list += tensor.tolist()
+
+                from itertools import groupby
+                global_recv_ranks_list.sort()
+                frequency = [len(list(group)) for key, group in groupby(global_recv_ranks_list)]
+                self.assertEqual(dist.get_world_size(), len(frequency))
+                self.assertEqual([2 * (dist.get_world_size() - 1)] * dist.get_world_size(), frequency)
+                self._barrier()
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/distributed_test.py:1063]: this barrier is only needed for the profiler enabled case?

**Review by mrshenli — APPROVED** [review]
Stamp to unblock.

File: torch/testing/_internal/distributed/distributed_test.py (hunk 2 above)
Comment (rohan-varma) [reply torch/testing/_internal/distributed/distributed_test.py:1063]: It looks like test_send_recv_nccl has this without profiling tests, hence thought we should keep it here as well. Although maybe that test copied the barrier from this test and it's not needed when profiler is not enabled.

**Pull request closed without merging**